File: THCS/2023_2024/Hue_2324_Phong/Bai3/BAI3.py
# Đọc dữ liệu từ file MUAHANG.INP
with open("MUAHANG.INP", "r") as infile:
    # Đọc số lượng hàng hóa và trọng lượng tối đa của xe đẩy
    N, M = map(int, infile.readline().split())
    
    # Lưu trữ trọng lượng và giá trị của mỗi hàng hóa
    weights = []
    values = []
    
    for _ in range(N):
        w, v = map(int, infile.readline().split())
        weights.append(w)
        values.append(v)

# Khởi tạo bảng dp
dp = [[0] * (M + 1) for _ in range(N + 1)]

# Lặp qua từng hàng hóa
for i in range(1, N + 1):
    for w in range(1, M + 1):
        if weights[i - 1] <= w:
            dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weights[i - 1]] + values[i - 1])
        else:
            dp[i][w] = dp[i - 1][w]

# Kết quả tối đa có thể đạt được
max_value = dp[N][M]

# Tìm danh sách các hàng hóa được chọn
chosen_items = []
w = M
for i in range(N, 0, -1):
    if dp[i][w] != dp[i - 1][w]:  # Nếu dp[i][w] khác dp[i-1][w] thì có nghĩa là đã chọn hàng hóa i
        chosen_items.append(i)
        w -= weights[i - 1]

chosen_items.sort()  # Sắp xếp lại thứ tự chỉ số từ nhỏ đến lớn
total_weight = sum(weights[i - 1] for i in chosen_items)  # Tính tổng trọng lượng

# Ghi kết quả ra file MUAHANG.OUT
with open("MUAHANG.OUT", "w") as outfile:
    outfile.write(f"{max_value}\n")
    outfile.write(" ".join(map(str, chosen_items)) + "\n")
    outfile.write(f"{total_weight}\n")

File: THCS/2023_2024/Hue_2324_Phong/Bai3/BAI3.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>

using namespace std;

int main() {
    ifstream infile("MUAHANG.INP");
    ofstream outfile("MUAHANG.OUT");

    int N, M;
    infile >> N >> M;

    vector<int> weights(N + 1);
    vector<int> values(N + 1);
    
    for (int i = 1; i <= N; ++i) {
        infile >> weights[i] >> values[i];
    }
    
    // Khởi tạo bảng dp
    vector<vector<int>> dp(N + 1, vector<int>(M + 1, 0));

    // Điền giá trị cho bảng dp
    for (int i = 1; i <= N; ++i) {
        for (int w = 1; w <= M; ++w) {
            if (weights[i] <= w) {
                dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weights[i]] + values[i]);
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }
    
    // Tìm giá trị lớn nhất có thể đạt được
    int max_value = dp[N][M];
    
    // Truy ngược lại các hàng hóa đã chọn
    vector<int> chosen_items;
    int w = M;
    for (int i = N; i >= 1; --i) {
        if (dp[i][w] != dp[i - 1][w]) {
            chosen_items.push_back(i);
            w -= weights[i];
        }
    }

    // Sắp xếp chỉ số các hàng hóa theo thứ tự từ nhỏ đến lớn
    sort(chosen_items.begin(), chosen_items.end());
    
    // Tính tổng trọng lượng của các hàng hóa đã chọn
    int total_weight = 0;
    for (int item : chosen_items) {
        total_weight += weights[item];
    }

    // Ghi kết quả ra file MUAHANG.OUT
    outfile << max_value << endl;
    for (size_t i = 0; i < chosen_items.size(); ++i) {
        if (i > 0) outfile << " ";
        outfile << chosen_items[i];
    }
    outfile << endl;
    outfile << total_weight << endl;

    return 0;
}
